### PR TITLE
module 2

### PR DESCRIPTION
**Pull request by dziyanabahdanava**

### CODE REVIEW

File: src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java [hunk 1]
```diff
@@ -55,4 +74,10 @@ private void copyProgramData(DefaultNutritionProgram from, DefaultNutritionProgr
         to.setCalories(from.getCalories());
         to.setNumberOfDays(from.getNumberOfDays());
     }
+
+    private void sendEventToQueue(String eventKey, DefaultNutritionProgram program) throws JsonProcessingException {
```
Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java]: usually on production we don't really throw Exception through all the layers
we wrap it in some RuntimeException (ServiceException, for instance)
```
try {
   template.convertAndSend(NUTRITION_EVENTS_QUEUE, jacksonMapper.writeValueAsString(event));
} catch (JsonProcessingException e) {
   throw new ServiceException(String.format("event %s could not be serialized", event), e);
}
```

Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java]: I'd also propose to have a dedicated class that will be responsible for working with the Queue.

File: src/main/java/com/epam/ms/config/RabbitConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.core.RabbitAdmin;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Slf4j
+@Configuration
+public class RabbitConfiguration {
+    private static final String HOST = "localhost";
```
Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/config/RabbitConfiguration.java]: It's a best practice to have all the configurations to be defined in `application.yml` file.
Then you use `@Value("${my.custom.property.path}") String customPath;` to access this configuration

File: build.gradle [hunk 3]
```diff
@@ -25,9 +25,12 @@ dependencies {
     testImplementation('org.springframework.boot:spring-boot-starter-test') {
         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
     }
-    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
-    implementation('org.postgresql:postgresql')
     implementation 'org.liquibase:liquibase-core'
+    implementation 'org.mongodb:mongo-java-driver'
+    implementation 'org.springframework.data:spring-data-mongodb'
+    implementation 'org.springframework.amqp:spring-rabbit'
+    compile group: 'com.google.guava', name: 'guava', version: '29.0-jre'
```
Comment (TsybulskyStepan) [inline build.gradle]: do you really need guava? :)

Comment (dziyanabahdanava) [reply build.gradle]: I use this dependency to get ImmutableMap

Comment (TsybulskyStepan) [reply build.gradle]: guava is really heavy dependency... it isn't worth having it only for ImmutableMap
Please, read once again about Java 9-14 features:
https://www.marcobehler.com/guides/a-guide-to-java-versions-and-features
you will find `Map.of()` function to be pretty similar to `ImmutableMap.of()`

File: build.gradle [hunk 4]
```diff
@@ -25,9 +25,14 @@ dependencies {
...
         exclude group: 'org.junit.vintage', module: 'junit-vintage-engine'
     }
-    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
-    implementation('org.postgresql:postgresql')
     implementation 'org.liquibase:liquibase-core'
+    implementation 'org.mongodb:mongo-java-driver'
+    implementation 'org.springframework.data:spring-data-mongodb'
+    implementation 'org.springframework.amqp:spring-rabbit'
+    runtime group: 'org.springframework.cloud', name: 'spring-cloud-dependencies', version: 'Hoxton.SR7', ext: 'pom'
+    //compile group: 'org.springframework.cloud', name: 'spring-cloud-starter-feign', version: '1.4.7.RELEASE'
```
Comment (TsybulskyStepan) [inline build.gradle]: why did you comment this line? 😄 
https://www.baeldung.com/spring-cloud-openfeign

File: build.gradle [hunk 5]
```diff
@@ -25,9 +25,14 @@ dependencies {
...
+    compile group: 'io.github.openfeign', name: 'feign-okhttp', version: '11.0'
```
Comment (TsybulskyStepan) [inline build.gradle]: I think if you had commented dependency you then you would not need these two feign-okhttp and feign-gson

File: src/main/java/com/epam/ms/user/UserProfileHandler.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        double factor = switch(profile.getGoal()) {
+            case SLIMMING, CUT, BODY_RELIEF -> 0.99;
+            case BULK -> 1.2;
+            case MAINTENANCE -> 1.08;
+        };
+        return (int) Math.round(calculateCalorieRate(profile) * factor);
+    }
+
+    private double calculateCalorieRate(UserProfile profile) {
+        double factor = switch(profile.getPhysicalActivity()) {
```
Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/user/UserProfileHandler.java:34]: wow, what is that? Java14? 😱

File: src/main/java/com/epam/ms/user/UserProfileHandler.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
+package com.epam.ms.user;
+
+import com.epam.ms.user.dto.UserProfile;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserProfileHandler {
+    private static final double WEIGHT_FACTOR = 9.99;
```
Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/user/UserProfileHandler.java]: also ,really convenient to have a property file for that
in spring boot they try to put EVERYTHING to config file. Then you can either use `@Value()` or! [`@ConfigurationProperties`](https://www.baeldung.com/configuration-properties-in-spring-boot) to inject a lot of configs at once.

File: src/main/java/com/epam/ms/user/UserProfileHandler.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package com.epam.ms.user;
```
Comment (TsybulskyStepan) [inline src/main/java/com/epam/ms/user/UserProfileHandler.java:1]: Let's try one more tool: Actuator. It's widely used in production
[Link](https://www.baeldung.com/spring-boot-actuators)

After that read [this](https://www.baeldung.com/spring-reloading-properties#reloading-cloud)
And try to play with it. Start app, try to call it. Find the results. Then go to your property files. Change values. Call /refresh endpoint. Call service one more time. Find new results.

File: src/main/java/com/epam/ms/user/UserProfileHandler.java (hunk 6 above)
Comment (dziyanabahdanava) [reply src/main/java/com/epam/ms/user/UserProfileHandler.java:34]: Yeeah😂

File: src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java (hunk 1 above)
Comment (dziyanabahdanava) [reply src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java]: Thank you, done

File: build.gradle (hunk 3 above)
Comment (dziyanabahdanava) [reply build.gradle]: 😳🙈

File: src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java (hunk 1 above)
Comment (dziyanabahdanava) [reply src/main/java/com/epam/ms/service/DefaultNutritionProgramService.java]: Thank you, done

File: src/main/java/com/epam/ms/config/RabbitConfiguration.java (hunk 2 above)
Comment (dziyanabahdanava) [reply src/main/java/com/epam/ms/config/RabbitConfiguration.java]: Thank you, done

File: src/main/java/com/epam/ms/user/UserProfileHandler.java (hunk 7 above)
Comment (dziyanabahdanava) [reply src/main/java/com/epam/ms/user/UserProfileHandler.java]: Thank you, done. Please review again, did I get the idea right?

**Pull request merged**